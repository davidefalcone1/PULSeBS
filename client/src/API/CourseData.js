//usata per la retrieve della lista dei corsi di un docente
class CourseData{
    constructor(courseId, courseName, teacherId, 
        normalBookingsAvgWeek, cancelledBookingsAvgWeek, waitingBookingsAvgWeek, 
        normalBookingsAvgMonth, cancelledBookingsAvgMonth, waitingBookingsAvgMonth,
        attendanceCountAvgMonth, attendanceCountAvgWeek ) {
        if (courseId) 
            this.courseId = courseId;
        this.courseName = courseName;
        this.teacherId = teacherId;
        this.normalBookingsAvgWeek = normalBookingsAvgWeek;
        this.cancelledBookingsAvgWeek = cancelledBookingsAvgWeek;
        this.waitingBookingsAvgWeek = waitingBookingsAvgWeek;
        this.normalBookingsAvgMonth = normalBookingsAvgMonth;
        this.cancelledBookingsAvgMonth = cancelledBookingsAvgMonth;
        this.waitingBookingsAvgMonth = waitingBookingsAvgMonth;
        this.attendanceCountAvgWeek = attendanceCountAvgWeek;
        this.attendanceCountAvgMonth = attendanceCountAvgMonth;
    }

    static fromJson(json){
        const temp =  Object.assign(new CourseData(), json);
        return temp;
    }
}

export default CourseData;