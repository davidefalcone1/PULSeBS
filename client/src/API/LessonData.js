import moment from 'moment';

class LessonData {
    constructor(scheduleId, courseId, startDate, endDate, occupiedSeats, availableSeats,
        isLessonCancelled, isLessonRemote, classroom
        /*, normalBookings, cancelledBookings, waitingBookings */) {
        if (scheduleId)
            this.scheduleId = scheduleId;
        this.courseId = courseId;
        this.startDate = moment(new Date(startDate));
        this.endDate = moment(new Date(endDate));
        this.occupiedSeats = occupiedSeats;
        this.availableSeats = availableSeats;
        this.isLessonCancelled = isLessonCancelled;
        this.isLessonRemote = isLessonRemote;
        this.classroom = classroom

        //TODO use the correct variables
        this.normalBookings = 0;
        this.cancelledBookings = 0;
        this.waitingBookings = 0;
    }

    static fromJson(json) {
        const temp = Object.assign(new LessonData(), json);
        temp.startDate = moment(new Date(temp.startDate));
        temp.endDate = moment(new Date(temp.endDate));
        return temp;
    }
}

export default LessonData;