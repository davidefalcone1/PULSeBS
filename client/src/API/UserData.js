class UserData{
    constructor(id, personId, fullName, email) {
        if (id) 
            this.id = id;
        this.personId = personId;
        this.fullName = fullName;
        this.email = email;
    }

    static fromJson(json){
        const temp =  Object.assign(new UserData(), json);
        return temp;
    }
}

export default UserData;