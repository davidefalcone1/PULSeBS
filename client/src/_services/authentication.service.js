import { BehaviorSubject } from 'rxjs';


const currentUserSubject = new BehaviorSubject(JSON.parse(localStorage.getItem('user')));

export const authenticationService = {
    login,
    logout,
    user: currentUserSubject.asObservable(),
    get currentUserValue () { return currentUserSubject.value }
};

function handleResponse(response) {
    return response.text().then(text => {
        const data = text && JSON.parse(text);
        if (!response.ok) {
            if ([401, 403].indexOf(response.status) !== -1) {
                // auto logout if 401 Unauthorized or 403 Forbidden response returned from api
                authenticationService.logout();
            }

            const error = (data && data.message) || response.statusText;
            return Promise.reject(error);
        }

        return data;
    });
}

function login(username, password) {
    const requestOptions = {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ username, password })
    };
    return new Promise(async function (resolve, reject) {
        try {
            const response = await fetch('/users/authenticate', {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                },
                body: JSON.stringify({ username, password }),
            });
            const user = await response.json().then(handleResponse);
            if (response.ok) {
                localStorage.setItem('user', JSON.stringify(user));
                currentUserSubject.next(user);
                resolve(user);
            }
            else
                reject();
        }
        catch (e) {
            reject();
        }
    });

}
function logout() {

    const requestOptions = {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
    };
    localStorage.removeItem('user');
    currentUserSubject.next(null);
    return fetch('/logout',requestOptions);
}