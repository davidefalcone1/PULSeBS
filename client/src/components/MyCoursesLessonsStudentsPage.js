import React from 'react';
import MyCourseLessonsStudentItem from './MyCoursesLessonsStudentItem';
import ListGroup from 'react-bootstrap/ListGroup';
import Accordion from 'react-bootstrap/Accordion';
import Card from 'react-bootstrap/Card';
import Button from 'react-bootstrap/Button'
import moment from 'moment'

const myCoursesLessons = (props) => {
  return (
    <MyCoursesLessonsPageRender teacherCourses={props.teacherCourses}
      myTeachedCoursesLessons={props.myTeachedCoursesLessons}
      studentsBookedToMyLessons={props.studentsBookedToMyLessons}
      myBookedStudentsInfos={props.myBookedStudentsInfos}
      cancelLesson={props.cancelLesson}
      changeLessonToRemote={props.changeLessonToRemote} />
  );
}

class MyCoursesLessonsPageRender extends React.Component {

  constructor(props) {
    super(props);
    this.props = props;
    this.state = {
      startingDay: undefined, endingDay: undefined
    }
  }

  render() {
    return (
      <>
        {this.props.teacherCourses && this.props.myTeachedCoursesLessons &&
          <div style={{ padding: "15px" }}>
            <Accordion>
              {this.props.teacherCourses.map((teacherCourse) => //per ogni mio corso
                <Card>
                  <Card.Header>
                    <Accordion.Toggle as={Button} variant="link" eventKey={"course-" + teacherCourse.courseName}>
                      <CourseHeader course={teacherCourse.courseName} />
                    </Accordion.Toggle>
                  </Card.Header>
                  <Accordion.Collapse eventKey={"course-" + teacherCourse.courseName}>
                    <Card.Body>
                      <Accordion>
                        {this.props.myTeachedCoursesLessons.map((courseLesson) => //per ogni lezione del mio corso
                          (courseLesson.courseId === teacherCourse.courseId) &&
                          <Card>
                            <Card.Header>
                              <div className="d-flex w-100 pt-3 justify-content-between no-gutters">
                                <Accordion.Toggle as={Button} variant="link" eventKey={teacherCourse.courseName + "-" + courseLesson.scheduleId}>
                                  <LessonHeader id={courseLesson.scheduleId} startingTime={courseLesson.startingTime} endingTime={courseLesson.endingTime}
                                    cancelLesson={this.props.cancelLesson} changeLessonToRemote={this.props.changeLessonToRemote} />
                                </Accordion.Toggle>
                                <LessonsHeaderButtons id={courseLesson.scheduleId} startingTime={courseLesson.startingTime} endingTime={courseLesson.endingTime}
                                  cancelLesson={this.props.cancelLesson} changeLessonToRemote={this.props.changeLessonToRemote}
                                  isLessonRemote={courseLesson.isLessonRemote}
                                  isLessonCancelled={courseLesson.isLessonCancelled} />
                              </div>
                            </Card.Header>
                            <Accordion.Collapse eventKey={teacherCourse.courseName + "-" + courseLesson.scheduleId}>
                              <Card.Body>
                                <StudentHeader />
                                <ListGroup as="ul" variant="flush">
                                  {this.props.studentsBookedToMyLessons.map((studentBooking) => //per ogni prenotazione alla lezione del mio corso
                                    (studentBooking.scheduleId === courseLesson.scheduleId) &&
                                    <ListGroup.Item key={studentBooking.studentId} id={teacherCourse.courseName + "-" + courseLesson.scheduleId + "-" + studentBooking.studentId}>
                                      {this.props.myBookedStudentsInfos.map((student) => //trova e mostra i dati dello studente
                                        (studentBooking.studentId === student.personId) &&
                                        <MyCourseLessonsStudentItem key={student.personId} student={student} />
                                      )}
                                    </ListGroup.Item>
                                  )}
                                </ListGroup>
                              </Card.Body>
                            </Accordion.Collapse>
                          </Card>
                        )}
                      </Accordion>
                    </Card.Body>
                  </Accordion.Collapse>
                </Card>
              )}
            </Accordion>
          </div>
        }

        {!this.props.teacherCourses && !this.props.myTeachedCoursesLessons &&
          <NoItemsImage />
        }
      </>
    )
  }
}

function CourseHeader(props) {
  return (
    <div className="d-flex w-100 pt-3 justify-content-between no-gutters" id={"course-" + props.course}>
      <h4>{props.course}</h4>
    </div>
  );
}
function LessonHeader(props) {
  return (
    <>
      <div id={"lesson-" + props.startingTime + "----" + props.endingTime}>
        <h6>Lezione del {props.startingTime.format("ddd DD-MM-YYYY HH:mm").toString()} -- {props.endingTime.format("ddd DD-MM-YYYY HH:mm").toString()}</h6>
      </div>
    </>
  );
}
function LessonsHeaderButtons(props) {
  return (
    <>
      <div id={"lesson-" + props.startingTime + "----" + props.endingTime}>
        {(moment().isBefore(moment(props.startingTime).subtract(30, 'm'))) && props.isLessonRemote &&
          <Button variant="warning" onClick={(event) => {
            event.preventDefault();
            props.changeLessonToRemote(props.id);
          }} id={"makeRemoteFieldOfLesson" + props.id}>
            Make Remote
          </Button>
        }
        <span>   </span>
        {(moment().isBefore(moment(props.startingTime).subtract(1, 'h'))) && props.isLessonCancelled &&
          <Button variant="danger" onClick={(event) => {
            event.preventDefault();
            props.cancelLesson(props.id);
          }} id={"cancelFieldOfLesson" + props.id}>
            CANCEL
          </Button>
        }
      </div>
    </>
  );
}

function StudentHeader(props) {
  return (
    <ListGroup.Item id={"ticketList-header"}>
      <div className="d-flex w-100 pt-3 justify-content-between no-gutters">
        <div className="col-sm-2">
          <h6>Person Id</h6>
        </div>
        <div className="col-sm-5">
          <h6>Full Name</h6>
        </div>
        <div className="col-sm-5">
          <h6>E-mail</h6>
        </div>
      </div>
    </ListGroup.Item>
  );
}

function NoItemsImage(props) {
  return (
    <div className="col-sm-12 pt-3">
      <img width="800" height="600" className="img-button" src='./images/no_data_image.png' alt="" />
    </div>
  );
}

export default myCoursesLessons;
