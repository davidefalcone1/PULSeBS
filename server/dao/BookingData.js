class BookingData {
    constructor(id, scheduleId, studentId, status, attended) {
        if (id)
            this.id = id;
        this.scheduleId = scheduleId;
        this.studentId = studentId;
        this.status = status;
        this.attended = attended;
    }

    static fromJson(json) {
        const temp = Object.assign(new BookingData(), json);
        return temp;
    }
}

module.exports = BookingData;