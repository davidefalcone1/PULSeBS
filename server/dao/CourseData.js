//usata per la retrieve della lista dei corsi
class CourseData{
    constructor(courseId, courseName, teacherId) {
        if (courseId) 
            this.courseId = courseId;
        this.courseName = courseName;
        this.teacherId = teacherId;
    }
}

module.exports = CourseData;