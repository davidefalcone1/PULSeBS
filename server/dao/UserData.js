class UserData{
    constructor(id, personId, fullName, email) {
        if (id) 
            this.id = id;
        this.personId = personId;
        this.fullName = fullName;
        this.email = email;
    }

}

module.exports = UserData;