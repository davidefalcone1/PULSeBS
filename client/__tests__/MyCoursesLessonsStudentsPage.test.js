import React from 'react';
import renderer from 'react-test-renderer';
import { AuthContext } from '../src/_services/AuthContext';
import MyCoursesLessonsStudentsPage from '../src/components/MyCoursesLessonsStudentsPage';
import { BrowserRouter } from 'react-router-dom';
import UserData from '../src/API/UserData';
import LessonData from '../src/API/LessonData';
import BookingData from '../src/API/BookingData';
import CourseData from '../src/API/CourseData';
import moment from 'moment';

Date.now = jest.fn(() => 1482363367071);//Avoid dynamic content on the page

describe('MyCoursesLessonsStudentsPage', ()=>{
    test('Renders correctly', ()=>{
        const component = renderer.create(
            <BrowserRouter>
                <AuthContext.Provider value={{user: new UserData(1, '12345', 'Davide Falcone', 'davide.falcone@polito.it')}}>
                    <MyCoursesLessonsStudentsPage teacherCourses={[new CourseData('XY123', 'Software engineering 2', '654321', 2, 3, 5, 6, 7, 3, 5, 7)]} myTeachedCoursesLessons={[new LessonData(1, 'XY123', moment().add(1, 'days'), moment().add(1, 'days').add(1, 'hours'), 9, 90, false, true, 'A1', 100, 4, 33, 0)]} studentsBookedToMyLessons={[new BookingData(1, 1, '12345', 1, 0)]} myBookedStudentsInfos={[new UserData(1, '12345', 'Davide Falcone', 'davide.falcone@polito.it')]}/>
                </AuthContext.Provider>
            </BrowserRouter>
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});