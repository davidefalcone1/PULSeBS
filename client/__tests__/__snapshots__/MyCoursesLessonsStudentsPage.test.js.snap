// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyCoursesLessonsStudentsPage Renders correctly 1`] = `
<div
  style={
    Object {
      "padding": "15px",
    }
  }
>
  <div
    className="accordion"
  >
    <div
      className="card"
    >
      <div
        className="card-header"
      >
        <button
          className="btn btn-link"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <div
            className="d-flex w-100 pt-3 justify-content-between no-gutters"
            id="course-Software engineering 2"
          >
            <h4>
              Software engineering 2
            </h4>
          </div>
        </button>
      </div>
      <div
        aria-expanded={null}
        className="collapse"
      >
        <div
          className="card-body"
        >
          <div
            className="accordion"
          >
            <div>
              <div
                className="row"
              >
                <div
                  className="col"
                >
                  <h5>
                    Statistics per week
                  </h5>
                  <h6>
                    10
                     total bookings, 
                     
                    2
                     actual bookings, 
                     
                    3
                     cancelled bookings, 
                     
                    5
                     waiting bookings
                     
                  </h6>
                </div>
                <div
                  className="col"
                >
                  <h5>
                    Statistics per month
                  </h5>
                  <h6>
                    16
                     total bookings, 
                     
                    6
                     actual bookings, 
                     
                    7
                     cancelled bookings,
                    3
                     waiting bookings
                  </h6>
                </div>
              </div>
              <br />
            </div>
            <div
              className="card"
            >
              <div
                className="card-header"
              >
                <div
                  className="d-flex w-100 pt-3 justify-content-between no-gutters"
                >
                  <button
                    className="btn btn-link"
                    disabled={false}
                    onClick={[Function]}
                    type="button"
                  >
                    <div
                      id="lesson-1482449767071----1482453367071"
                    >
                      <h6>
                        Lesson of 
                        Fri 23-12-2016 00:36
                         -- 
                        Fri 23-12-2016 01:36
                         
                        <span
                          className="badge badge-pill badge-warning"
                        >
                          Remote
                        </span>
                         
                      </h6>
                    </div>
                  </button>
                  <div
                    id="1buttons"
                  >
                    <span>
                         
                    </span>
                    <button
                      className="btn btn-danger"
                      disabled={false}
                      id="cancelFieldOfLesson1"
                      onClick={[Function]}
                      type="button"
                    >
                      CANCEL
                    </button>
                  </div>
                </div>
              </div>
              <div
                aria-expanded={null}
                className="collapse"
              >
                <div
                  className="card-body"
                >
                  <h6>
                    137
                     total bookings, 
                     
                    100
                     actual bookings, 
                     
                    4
                     cancelled bookings, 
                     
                    33
                     waiting bookings, 
                     
                    0
                     attendances
                  </h6>
                  <br />
                  <div
                    className="list-group-item"
                    disabled={false}
                    id="ticketList-header"
                    onClick={[Function]}
                  >
                    <div
                      className="d-flex w-100 pt-3 justify-content-between no-gutters"
                    >
                      <div
                        className="col-sm-2"
                      >
                        <h6>
                          Person Id
                        </h6>
                      </div>
                      <div
                        className="col-sm-5"
                      >
                        <h6>
                          Full Name
                        </h6>
                      </div>
                      <div
                        className="col-sm-5"
                      >
                        <h6>
                          E-mail
                        </h6>
                      </div>
                    </div>
                  </div>
                  <ul
                    className="list-group list-group-flush"
                    onKeyDown={[Function]}
                  >
                    <div
                      className="list-group-item"
                      data-rb-event-key={null}
                      disabled={false}
                      id="Software engineering 2-1-12345"
                      onClick={[Function]}
                    >
                      <div
                        className="list-group-item"
                        data-rb-event-key={null}
                        disabled={false}
                        id="student-12345"
                        onClick={[Function]}
                      >
                        <div
                          className="d-flex w-100 pt-3 justify-content-between no-gutters"
                        >
                          <div
                            className="col-sm-2"
                          >
                            <p
                              id="idOfStudent_12345"
                            >
                              12345
                            </p>
                          </div>
                          <div
                            className="col-sm-3"
                          >
                            <p
                              id="fullNameOfStudent_12345"
                            >
                              Davide Falcone
                            </p>
                          </div>
                          <div
                            className="col-sm-4"
                          >
                            <p
                              id="emailOfStudent12345"
                            >
                              davide.falcone@polito.it
                            </p>
                          </div>
                          <div
                            className="col-sm-3"
                          >
                            <button
                              className="btn btn-success"
                              disabled={false}
                              id="setPresentOf1 - 12345"
                              onClick={[Function]}
                              type="button"
                            >
                              Set as Present
                            </button>
                            0
                          </div>
                        </div>
                      </div>
                    </div>
                  </ul>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
